

## Problem Statement

Use Multiple Linear Regression to **predict the consumption of petrol** given relevant variables are the petrol tax, the per capita, income, the number of miles of paved highway, and the proportion of the population with driver's licenses.


## Question 1 - Exploratory Data Analysis

*Read the dataset given in file named **'petrol.csv'**. Check the statistical details of the dataset.*

**Hint:** You can use **df.describe()**

In [1]:
import numpy as np
import pandas as pd
df=pd.read_csv('petrol.csv')
df.head()

Petrol_tax  Average_income  Paved_Highways  Population_Driver_licence(%)  \
0         9.0            3571            1976                         0.525   
1         9.0            4092            1250                         0.572   
2         9.0            3865            1586                         0.580   
3         7.5            4870            2351                         0.529   
4         8.0            4399             431                         0.544   

   Petrol_Consumption  
0                 541  
1                 524  
2                 561  
3                 414  
4                 410

In [2]:
df1=df
df.shape

(48, 5)

In [3]:
# check information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48 entries, 0 to 47
Data columns (total 5 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Petrol_tax                    48 non-null     float64
 1   Average_income                48 non-null     int64  
 2   Paved_Highways                48 non-null     int64  
 3   Population_Driver_licence(%)  48 non-null     float64
 4   Petrol_Consumption            48 non-null     int64  
dtypes: float64(2), int64(3)
memory usage: 2.0 KB


In [4]:
# check null values   - No null values
df.isnull().sum()

Petrol_tax                      0
Average_income                  0
Paved_Highways                  0
Population_Driver_licence(%)    0
Petrol_Consumption              0
dtype: int64

In [5]:
# check duplpicates - No duplicates found
df[df.duplicated()]

Empty DataFrame
Columns: [Petrol_tax, Average_income, Paved_Highways, Population_Driver_licence(%), Petrol_Consumption]
Index: []

In [7]:
# check statistical information
df.describe([0.99])

Petrol_tax  Average_income  Paved_Highways  \
count   48.000000       48.000000       48.000000   
mean     7.668333     4241.833333     5565.416667   
std      0.950770      573.623768     3491.507166   
min      5.000000     3063.000000      431.000000   
50%      7.500000     4298.000000     4735.500000   
99%      9.530000     5331.190000    16091.880000   
max     10.000000     5342.000000    17782.000000   

       Population_Driver_licence(%)  Petrol_Consumption  
count                     48.000000           48.000000  
mean                       0.570333          576.770833  
std                        0.055470          111.885816  
min                        0.451000          344.000000  
50%                        0.564500          568.500000  
99%                        0.701910          919.590000  
max                        0.724000          968.000000

# Question 2 - Cap outliers 

Find the outliers and cap them. (Use (Q1 - 1.5 * IQR) as the minimum cap and (Q3 + 1.5 * IQR) as the max cap. The decision criteria is you should consider the datapoints which only falls within this range. The data points which fall outside this range are outliers and the entire row needs to be removed

In [8]:
# Petrol_tax
q1=df1['Petrol_tax'].quantile(0.25)
q3=df1['Petrol_tax'].quantile(0.75)
iqr=q3-q1
print(iqr)
ll=q1-1.5*iqr
ul=q3+1.5*iqr
df1=df1[~((df['Petrol_tax']<ll) | (df1['Petrol_tax']>ul))]
df1.shape

1.125


(46, 5)

In [9]:
# Average_income
q1=df1['Average_income'].quantile(0.25)
q3=df1['Average_income'].quantile(0.75)
iqr=q3-q1
print(iqr)
ll=q1-1.5*iqr
ul=q3+1.5*iqr
df1=df1[~((df1['Average_income']<ll) | (df1['Average_income']>ul))]
df1.shape

831.5


(46, 5)

In [10]:
# Paved_Highways
q1=df1['Paved_Highways'].quantile(0.25)
q3=df1['Paved_Highways'].quantile(0.75)
iqr=q3-q1
print(iqr)
ll=q1-1.5*iqr
ul=q3+1.5*iqr
df1=df1[~((df1['Paved_Highways']<ll) | (df1['Paved_Highways']>ul))]
df1.shape

3594.5


(45, 5)

In [11]:
# Population_Driver_licence
q1=df1['Population_Driver_licence(%)'].quantile(0.25)
q3=df1['Population_Driver_licence(%)'].quantile(0.75)
iqr=q3-q1
print(iqr)
ll=q1-1.5*iqr
ul=q3+1.5*iqr
df1=df1[~((df1['Population_Driver_licence(%)']<ll) | (df1['Population_Driver_licence(%)']>ul))]
df1.shape

0.07199999999999995


(44, 5)

In [12]:
# Petrol_Consumption
q1=df1['Petrol_Consumption'].quantile(0.25)
q3=df1['Petrol_Consumption'].quantile(0.75)
iqr=q3-q1
print(iqr)
ll=q1-1.5*iqr
ul=q3+1.5*iqr
df1=df1[~((df1['Petrol_Consumption']<ll) | (df1['Petrol_Consumption']>ul))]
df1.shape

110.75


(42, 5)

# Question 3 - Transform the dataset 
Divide the data into feature(X) and target(Y) sets.

In [13]:
df.corr()

Petrol_tax  Average_income  Paved_Highways  \
Petrol_tax                      1.000000        0.012665       -0.522130   
Average_income                  0.012665        1.000000        0.050163   
Paved_Highways                 -0.522130        0.050163        1.000000   
Population_Driver_licence(%)   -0.288037        0.157070       -0.064129   
Petrol_Consumption             -0.451280       -0.244862        0.019042   

                              Population_Driver_licence(%)  Petrol_Consumption  
Petrol_tax                                       -0.288037           -0.451280  
Average_income                                    0.157070           -0.244862  
Paved_Highways                                   -0.064129            0.019042  
Population_Driver_licence(%)                      1.000000            0.698965  
Petrol_Consumption                                0.698965            1.000000

In [14]:
df1.corr()

Petrol_tax  Average_income  Paved_Highways  \
Petrol_tax                      1.000000       -0.133841       -0.443926   
Average_income                 -0.133841        1.000000       -0.076862   
Paved_Highways                 -0.443926       -0.076862        1.000000   
Population_Driver_licence(%)   -0.316342        0.296002        0.133983   
Petrol_Consumption             -0.463247       -0.254464        0.215182   

                              Population_Driver_licence(%)  Petrol_Consumption  
Petrol_tax                                       -0.316342           -0.463247  
Average_income                                    0.296002           -0.254464  
Paved_Highways                                    0.133983            0.215182  
Population_Driver_licence(%)                      1.000000            0.549161  
Petrol_Consumption                                0.549161            1.000000

In [15]:
Y=df1.Petrol_Consumption
X=df1.drop(['Petrol_Consumption'],axis=1)

# Question 4 - Split data into train, test sets 
Divide the data into training and test sets with 80-20 split using scikit-learn. Print the shapes of training and test feature sets.

In [16]:
from sklearn.model_selection import train_test_split

In [30]:
Y=df1.Petrol_Consumption
X=df1.Petrol_tax
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.3,random_state =0)
print(X_train.shape,Y_train.shape)
print(X_test.shape,Y_test.shape)

(29,) (29,)
(13,) (13,)


# Question 5 - Build Model 
Estimate the coefficients for each input feature. Construct and display a dataframe with coefficients and X.columns as columns

In [31]:
import statsmodels.formula.api as smf
import statsmodels.api as sm
model=sm.OLS(Y_train,X_train).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:     Petrol_Consumption   R-squared (uncentered):                   0.955
Model:                            OLS   Adj. R-squared (uncentered):              0.953
Method:                 Least Squares   F-statistic:                              595.6
Date:                Fri, 13 May 2022   Prob (F-statistic):                    2.07e-20
Time:                        12:37:03   Log-Likelihood:                         -181.00
No. Observations:                  29   AIC:                                      364.0
Df Residuals:                      28   BIC:                                      365.4
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Petrol_tax    73.7599      3.022     24.405      0.000      67.569      79.951
==============================================================================
Omnibus:                        1.831   Durbin-Watson:                   1.587
Prob(Omnibus):                  0.400   Jarque-Bera (JB):                1.312
Skew:                           0.519   Prob(JB):                        0.519
Kurtosis:                       2.902   Cond. No.                         1.00
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# R-Square 

# Question 6 - Evaluate the model 
Calculate the accuracy score for the above model.

In [32]:
def mean_abs_per_err(pred,actual):
    per=(abs(actual-pred))/actual
    value=per*100
    return value

#### Accuracy of training model

In [33]:
pred_y=model.predict(X_train)
from sklearn import metrics
MSE=metrics.mean_squared_error(pred_y,Y_train)
print('Mean Squared Error:',MSE)
RMSE=np.sqrt(MSE)
print('Root Mean Squared Error:',RMSE)
MAPE=(mean_abs_per_err(pred_y,Y_train).sum())/Y_train.size
print("Mean Absolute Percentage Error:", MAPE)

Mean Squared Error: 15442.552329703223
Root Mean Squared Error: 124.26806641170218
Mean Absolute Percentage Error: 16.946606056874497


#### Accracy of testing model

In [34]:
pred_y=model.predict(X_test)
from sklearn import metrics
MSE=metrics.mean_squared_error(pred_y,Y_test)
print('Mean Squared Error:',MSE)
RMSE=np.sqrt(MSE)
print('Root Mean Squared Error:',RMSE)
MAPE=(mean_abs_per_err(pred_y,Y_test).sum())/Y_test.size
print("Mean Absolute Percentage Error:", MAPE)

Mean Squared Error: 12716.092228130798
Root Mean Squared Error: 112.7656518099851
Mean Absolute Percentage Error: 18.679494581809873


# Question 7: Repeat the same Multi linear regression modelling by adding both Income and Highway features
Find R2 


In [35]:
Y=df1.Petrol_Consumption
X=df1.drop(['Petrol_Consumption'],axis=1)
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.3,random_state =0)
print(X_train.shape,Y_train.shape)
print(X_test.shape,Y_test.shape)

(29, 4) (29,)
(13, 4) (13,)


In [36]:
model=sm.OLS(Y_train,X_train).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:     Petrol_Consumption   R-squared (uncentered):                   0.990
Model:                            OLS   Adj. R-squared (uncentered):              0.988
Method:                 Least Squares   F-statistic:                              599.9
Date:                Fri, 13 May 2022   Prob (F-statistic):                    1.96e-24
Time:                        12:37:10   Log-Likelihood:                         -159.67
No. Observations:                  29   AIC:                                      327.3
Df Residuals:                      25   BIC:                                      332.8
Df Model:                           4                                                  
Covariance Type:            nonrobust                                                  
================================================================================================
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------
Petrol_tax                      -5.4438     10.307     -0.528      0.602     -26.671      15.784
Average_income                  -0.0545      0.022     -2.481      0.020      -0.100      -0.009
Paved_Highways                   0.0028      0.004      0.639      0.529      -0.006       0.012
Population_Driver_licence(%)  1458.3297    197.498      7.384      0.000    1051.575    1865.084
==============================================================================
Omnibus:                        3.489   Durbin-Watson:                   1.476
Prob(Omnibus):                  0.175   Jarque-Bera (JB):                2.806
Skew:                           0.759   Prob(JB):                        0.246
Kurtosis:                       2.868   Cond. No.                     1.17e+05
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[3] The condition number is large, 1.17e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Training Accuracy

In [37]:
pred_y=model.predict(X_train)
from sklearn import metrics
MSE=metrics.mean_squared_error(pred_y,Y_train)
print('Mean Squared Error:',MSE)
RMSE=np.sqrt(MSE)
print('Root Mean Squared Error:',RMSE)
MAPE=(mean_abs_per_err(pred_y,Y_train).sum())/Y_train.size
print("Mean Absolute Percentage Error:", MAPE)

Mean Squared Error: 3546.304711741208
Root Mean Squared Error: 59.550858194833836
Mean Absolute Percentage Error: 8.105078070588407


#### Testing Accuracy

In [38]:
pred_y=model.predict(X_test)
from sklearn import metrics
MSE=metrics.mean_squared_error(pred_y,Y_test)
print('Mean Squared Error:',MSE)
RMSE=np.sqrt(MSE)
print('Root Mean Squared Error:',RMSE)
MAPE=(mean_abs_per_err(pred_y,Y_test).sum())/Y_test.size
print("Mean Absolute Percentage Error:", MAPE)

Mean Squared Error: 3159.4870155895214
Root Mean Squared Error: 56.20931431346162
Mean Absolute Percentage Error: 9.204626141430504


# Question 8: Print the coefficients of the multilinear regression model

In [39]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, Y_train)

LinearRegression()

In [40]:
regressor.coef_

array([-4.92778425e+01, -9.14403325e-02, -4.67258091e-03,  8.89320765e+02])

# Question 9 
In one or two sentences give reasoning on R-Square on the basis of above findings
Answer

When there was only one independent variable Petrol_tax in our regression the  R2  value was 95%
Adjusted  R2  value was 95.3%

When there were three independent variable Petrol_tax,Average_income,Paved_Highways and Population_Driver_licence(%) in our regression the  R2  value was 99%
Adjusted  R2  value was 98.8%

On addition of other varibles the  R2  value increased by a considerable amount of 4%. However the Adjusted  R2  value did not increase as much as  R2 . This is because Adjusted  R2  only increases if the variable contributes in the prediction on dependent variable